### ISSUE TITLE
Disallowing casting `Unknown` in variable definition in strict mode

### ISSUE THREAD

**Issue by math-fehr** [microsoft/pyright#10214]
**Feature requested**
When using strict mode, a generic type `A[Unknown]` should not be allowed to be implicitly casted to a `A[int]` type with no type error. Here is an example (tested using strict mode in version 1.1.398):

```python
from typing import Generic, TypeVar, Any

T = TypeVar("T")


class A(Generic[T]):
    field: T


def foo(x: Any):
    if isinstance(x, A):
        # Here, `x` is of type `A[Unknown]`
        y: A[int] = x  # Expected error here, as we are using subtyping of `Unknown`
        
        # Expected way of doing such a cast in strict mode
        # y: A[int] = cast(x, A[int])
```

The lack of error seems to be against the philosophy of strict mode with `Unknown`s, as almost any use of `Unknown` is triggering an error in pyright.

Looking at the [diagnostic settings](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#diagnostic-settings-defaults), it seems that none of them apply here, so this is more a feature request than a bug I think.

**Comment by erictraut** [comment]
I agree that the `reportUnknownVariableType` is a bit inconsistent here. 

I've updated the logic to use the pre-narrowed (assigned) type rather than the narrowed type to determine whether an unknown type should be reported. This change will create some noise for some pyright users who have enabled this check, but I think it's worth it for consistency.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.399.
